### DISCUSSION TITLE
Can RefindPlus Load External Drivers like rEFInd Does?

### DISCUSSION THREAD

**Discussion by 1457384613gh** [dakanji/RefindPlus#146] [General]
![image](https://user-images.githubusercontent.com/69227436/213923377-b9adc3af-2d99-4031-8975-ceecfeadcd37.png)

It shows that rEFInd can load [EFI network driver]-s and [EFI video driver]-s.
How about RefindPlus?
[efi_video.zip](https://github.com/dakanji/RefindPlus/files/10474713/efi_video.zip)
[CsmVideoDxe.efi] is from Clover; [BiosVideo.efi] is from OpenCore.
Could RefindPlus load one of them?
At least one of them is adapted from [IntelGopDriver.efi].

![image](https://user-images.githubusercontent.com/69227436/213923710-120c5a02-30ea-4005-b2fe-c8e9adc7b6d7.png)

How to make it support more UEFI resolutions?
Could RefindPlus load an external [IntelGopDriver.efi]?

**Comment by startergo** [comment]
Put it into the drivers folder and reboot.

**Comment by dakanji** [comment]
@1457384613gh ... Please add the BiosVideo.efi file from OpenCore (or the other one from Clover) to your drivers folder and confirm whether you get the resolution you are looking for with the attached build (x288).
[X288-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/11322415/X288-BOOTx64.zip)

Thanks

**Reply by 1457384613gh** [reply]
I haven't tried, yet.
Have you adapted one for loading those EFI_Video_Driver ?

**Reply by dakanji** [reply]
Please add the driver, test with the attached RefindPlus build and report back as was written.
The drivers would not have worked before due to driver priority issues which the build attempts to resolve.

**Reply by 1457384613gh** [reply]
I finally find that DUET can help resolutions of all, such as `Metro Windows Boot Manager`, `Grub2 with theme`, `rEFInd` and so on.

BiosVideoDXE.efi still cannot help at least resolutions of `Metro Windows Boot Manager` and `Grub2 with theme`.


I will try later.

**Reply by 1457384613gh** [reply]
No 1366x768
Still 1024x768

[23h25p4451.log](https://github.com/dakanji/RefindPlus/files/11323178/23h25p4451.log)
[23h25p4745.log](https://github.com/dakanji/RefindPlus/files/11323179/23h25p4745.log)

**Comment by dakanji** [comment]
Try again and press the 'ESC' key after it loads.

**Reply by 1457384613gh** [reply]
Above all, X288 runs without mouse, so I have to select an entry by using keyboard.
What you want to know is that the keyboard cannot work after pressing `ESC` so it cannot select any entries so I have to press `power` to restart.

[23h25q2210.log](https://github.com/dakanji/RefindPlus/files/11323538/23h25q2210.log)
No 1366x768
Still 1024x768

**Comment by dakanji** [comment]
> X288 runs without mouse, so I have to select an entry by using keyboard.

Refer to the ReadMe of this repo on how to enable mouse in RefindPlus

> the keyboard cannot work after pressing ESC so it cannot select any entries so I have to press power to restart.

I noticed that it hung. Will look at that one separately.

> Still 1024x768

Yeah. Doesn't work indeed.
Not sure whether it is because the BIosVideo driver seems to be UGA related and your GPU is GOP based.
Took a quick look at the source and everything there is about UGA.  Try to enable `pass_uga_through` as well as `provide_console_gop`

**Comment by dakanji** [comment]
You can give X289 a try.
[X289-BOOTx64.zip](https://github.com/dakanji/RefindPlus/files/11333096/X289-BOOTx64.zip)

The zip package includes a generic BiosVideo driver file as the one from OpenCore is tied to that project.
Try the ESC key if not initially working to be sure the driver is loaded.

Thanks

**Reply by 1457384613gh** [reply]
[23h27b5446.log](https://github.com/dakanji/RefindPlus/files/11338845/23h27b5446.log)
Stilllllllllllllllllllllllllllllllllllllllllll.

DUET can help more.
I think you can adapt DUET from theirs.
https://github.com/CloverHackyColor/CloverBootloader/tree/master/CloverEFI/BootSector
https://github.com/acidanthera/OpenCorePkg/tree/master/Legacy/BootSector

If DUET, all drivers of Legacy_BIOS can be used.

I am using DUET from Clover.
`MBR=>PBR=>boot=>CLOVERX64.efi (refind_x64.efi renamed)`
If you make your DUET, you can set
`MBR=>PBR=>boot=>bootx64.efi`
or
`MBR=>PBR=>boot=>refindplus.efi`

**Comment by dakanji** [comment]
I see. Best stick with Duet or OpenDuet as there is no plan to add booting RefindPlus via legacy bootcode (unless someone else does it) at this time.

Thanks
